runtime/gc_vlhgc/AllocationContextBalanced.cpp

```diff
@@ -123,9 +123,9 @@ MM_AllocationContextBalanced::flushInternal(MM_EnvironmentBase *env)
 {
 	/* flush all the regions we own for active allocation */
 	if (NULL != _allocationRegion){
-		MM_MemoryPoolBumpPointer *pool = (MM_MemoryPoolBumpPointer*)_allocationRegion->getMemoryPool();
+		MM_MemoryPoolAddressOrderedList *pool = (MM_MemoryPoolAddressOrderedList*)_allocationRegion->getMemoryPool();
```

Comment:
We have this pattern of fetching MP and casting the type to MPAOL many times.

In some cases it's not necessary. The type of pool variable can be base MP (and avoid casting), since the methods you later apply to them are virtual and exist in both MP and MPAOL.

For example, getActualFreeMemorySize() that you use here already exists in base MP, but recalculateMemoryPoolStatistics() does not, which unfortunately prevents you to simplify things.

But I don't see a problem of introducing recalculateMemoryPoolStatistics() in base MP (there is nothing in its name suggesting AOL specific).

So let's go through all the code and see if we can simplify things and how many of new methods we need to introduce to base MP. Of course it will be another OMR change before we can proceed with this one.